### PR TITLE
Implement Doppelganger Check

### LINKED ISSUE #7985: Check on-chain for attestations at validator launch
# 🚀 Feature Request

### Description
In the event of having a corrupted `validator.db` an easy and cheap safeguard would be to grab attestations from the last 3 epochs from the beacon-node and check if the validating indexes appear there. This would prevent having a double attestation in most cases. This test can be run only at launchtime and if the validator.db does not show any activity for the last/current epoch. So that the impact will be minimal in long running validators. 

I'd like to see what the team think of possible design patterns before submitting a PR, but I think this can be as easy as pulling the last 3 epochs blocks from the node, and comparing the attestations for the indexes of the validating accounts. And stopping if something is found on-chain that is not found on the `validator.db` with a possible flag `--disable-launchtime-check` or whatever to override this behavior. 

### Additional Info

Discussion in Discord: https://discord.com/channels/476244492043812875/730447046930071672/782041664327909417

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Feature Implementation

**What does this PR do? Why is it needed?**

- [x] Checks for validator balances in the last N epochs, and determines if there is a positive change.
- [x] Add tests for the feature.
- [x] Gate it behind a flag.

**Which issues(s) does this PR fix?**

Fixes #7985 

**Other notes for review**

### CODE REVIEW

File: beacon-chain/rpc/validator/status.go [hunk 1]
```diff
@@ -88,6 +88,90 @@ func (vs *Server) MultipleValidatorStatus(
...
+func (vs *Server) CheckDoppelGanger(ctx context.Context, req *ethpb.DoppelGangerRequest) (*ethpb.DoppelGangerResponse, error) {
+	if vs.SyncChecker.Syncing() {
+		return nil, status.Errorf(codes.Unavailable, "Syncing to latest head, not ready to respond")
+	}
+	headState, err := vs.HeadFetcher.HeadState(ctx)
+	if err != nil {
+		return nil, status.Error(codes.Internal, "Could not get head state")
+	}
+	// We walk back from the current head to the previous 2 states. With S_i , i := 0,1,2. i = 0
+	// would signify the current head state in this epoch.
```
Comment (potuz) [inline beacon-chain/rpc/validator/status.go:100]:
```suggestion
	// We walk back from the current head state to the state at the beginning of the previous 2 epochs. 
	// Where S_i , i := 0,1,2. i = 0 would signify the current head state in this epoch.
```

File: beacon-chain/rpc/validator/status.go [hunk 2]
```diff
@@ -88,6 +88,90 @@ func (vs *Server) MultipleValidatorStatus(
...
+		return nil, status.Error(codes.Internal, "Could not get previous state")
+	}
+	resp := &ethpb.DoppelGangerResponse{
+		Responses: []*ethpb.DoppelGangerResponse_ValidatorResponse{},
+	}
+	for _, v := range req.ValidatorRequests {
+		// If the validator's last recorded epoch was
+		// less than 2 epoch ago, this method will not
+		// be able to catch duplicates.
+		if v.Epoch+2 > currEpoch {
```
Comment (potuz) [inline beacon-chain/rpc/validator/status.go:125]:
```suggestion
		if v.Epoch+2 >= currEpoch {
```
Rationale: if the validator attested in `olderEpoch` and the attestation was included in `previousEpoch` the balance change will not be reflected until the end of `currentEpoch`.

**Review by potuz — CHANGES_REQUESTED** [review]
Haven't checked the syntax of the PR, but the design of the protection looks good to me, simple to read and effective.

File: beacon-chain/rpc/validator/status.go [hunk 3]
```diff
@@ -88,6 +88,101 @@ func (vs *Server) MultipleValidatorStatus(
...
+	if err != nil {
+		olderEpoch = previousEpoch
+	}
+	prevState, err := vs.StateGen.StateBySlot(ctx, params.BeaconConfig().SlotsPerEpoch.Mul(uint64(previousEpoch)))
+	if err != nil {
+		return nil, status.Error(codes.Internal, "Could not get previous state")
+	}
+	olderState, err := vs.StateGen.StateBySlot(ctx, params.BeaconConfig().SlotsPerEpoch.Mul(uint64(olderEpoch)))
+	if err != nil {
+		return nil, status.Error(codes.Internal, "Could not get previous state")
```
Comment (jmozah) [inline beacon-chain/rpc/validator/status.go:122]: "Could not get older state" would be appropriate.

File: beacon-chain/rpc/validator/server.go [hunk 4]
```diff
@@ -62,7 +62,7 @@ type Server struct {
 	Eth1BlockFetcher       powchain.POWBlockFetcher
 	PendingDepositsFetcher depositcache.PendingDepositsFetcher
 	OperationNotifier      opfeed.Notifier
-	StateGen               *stategen.State
+	StateGen               stategen.StateManager
```
Comment (ahadda5) [inline beacon-chain/rpc/validator/server.go:65]: For my benefit. why this change?

Comment (nisdas) [reply beacon-chain/rpc/validator/server.go:65]: Simply changes it from a struct to an interface, this allows us to mock it correctly for tests. `*stategen.State` fulfills the state manager interface.

**Comment by yorickdowne** [comment]
`config.go` and `flags.go` misspell "doppelganger" as "doppleganger". Correct spelling is "doppelganger" (well, anglicized, never mind the lack of umlauts :)).

File: beacon-chain/rpc/validator/status.go [hunk 5]
```diff
@@ -88,6 +88,101 @@ func (vs *Server) MultipleValidatorStatus(
...
+		// less than 2 epoch ago, this method will not
+		// be able to catch duplicates.
+		if v.Epoch+2 >= currEpoch {
+			resp.Responses = append(resp.Responses,
+				&ethpb.DoppelGangerResponse_ValidatorResponse{
+					PublicKey:       v.PublicKey,
+					DuplicateExists: false,
+				})
+			continue
+		}
```
Comment (prestonvanloon) [inline beacon-chain/rpc/validator/status.go:138]: Why is this the case? It isn't clear to the reader why a request within the last 2 epochs will be marked as false or safe to proceed without actually checking.

File: beacon-chain/rpc/validator/status.go [hunk 6]
```diff
@@ -88,6 +88,101 @@ func (vs *Server) MultipleValidatorStatus(
...
+	if err != nil {
+		return nil, status.Error(codes.Internal, "Could not get head state")
+	}
+	// We walk back from the current head state to the state at the beginning of the previous 2 epochs.
+	// Where S_i , i := 0,1,2. i = 0 would signify the current head state in this epoch.
+	currEpoch := helpers.SlotToEpoch(headState.Slot())
+	previousEpoch, err := currEpoch.SafeSub(1)
+	if err != nil {
+		previousEpoch = currEpoch
+	}
```
Comment (prestonvanloon) [inline beacon-chain/rpc/validator/status.go:111]: Why would this error? 
Please log something and do not ignore this error.
```suggestion
	if err != nil {
		log.WithError(err).Error("Could not determine previous epoch")	
		previousEpoch = currEpoch
	}
```

File: beacon-chain/rpc/validator/status.go [hunk 7]
```diff
@@ -88,6 +88,101 @@ func (vs *Server) MultipleValidatorStatus(
...
+	olderEpoch, err := previousEpoch.SafeSub(1)
+	if err != nil {
+		olderEpoch = previousEpoch
+	}
```
Comment (prestonvanloon) [inline beacon-chain/rpc/validator/status.go:115]:
```suggestion
	if err != nil {
		log.WithError(err).Error("Could not determine older epoch")	
		olderEpoch = previousEpoch
	}
```

File: beacon-chain/rpc/validator/status.go [hunk 8]
```diff
@@ -88,6 +88,101 @@ func (vs *Server) MultipleValidatorStatus(
...
+		if err != nil {
+			return nil, status.Error(codes.Internal, "Could not get validator's balance")
+		}
+		nextBal, err := prevState.BalanceAtIndex(valIndex)
+		if err != nil {
+			return nil, status.Error(codes.Internal, "Could not get validator's balance")
+		}
+		// If the next epoch's balance is higher, we mark it as an existing
+		// duplicate.
+		if nextBal > baseBal {
```
Comment (prestonvanloon) [inline beacon-chain/rpc/validator/status.go:154]: This is not a safe doppelganger detection. If the validator produced an attestation for a fork, it could be a decrease in balances. Is this the same implementation that other clients have designed?

File: shared/featureconfig/flags.go [hunk 9]
```diff
@@ -122,6 +122,10 @@ var (
 		Name:  "enable-optimized-balance-update",
 		Usage: "Enables the optimized method of updating validator balances.",
 	}
+	enableDoppleGangerProtection = &cli.BoolFlag{
+		Name:  "enable-doppleganger",
+		Usage: "Enables the validator to perform a doppleganger check.",
```
Comment (prestonvanloon) [inline shared/featureconfig/flags.go:127]: Please add some kind of warning. This is not a 100% effective check and the user should not rely on this feature to allow them to run in unsafe configurations.

File: shared/featureconfig/flags.go [hunk 10]
```diff
@@ -122,6 +122,10 @@ var (
 		Name:  "enable-optimized-balance-update",
 		Usage: "Enables the optimized method of updating validator balances.",
 	}
+	enableDoppleGangerProtection = &cli.BoolFlag{
```
Comment (prestonvanloon) [inline shared/featureconfig/flags.go:125]: Typo with Dopple

File: validator/client/validator.go [hunk 11]
```diff
@@ -369,6 +370,85 @@ func (v *validator) SlotDeadline(slot types.Slot) time.Time {
...
+			})
+	}
+	resp, err := v.validatorClient.CheckDoppelGanger(ctx, req)
+	if err != nil {
+		return err
+	}
+	// If nothing is returned by the beacon node, we can exit early.
+	if resp == nil || resp.Responses == nil || len(resp.Responses) == 0 {
+		return nil
+	}
```
Comment (prestonvanloon) [inline validator/client/validator.go:420]: This is a really weird response. Please log a warning or something.

File: validator/client/validator.go [hunk 12]
```diff
@@ -369,6 +370,85 @@ func (v *validator) SlotDeadline(slot types.Slot) time.Time {
...
+	for _, valRes := range resp.Responses {
+		if valRes.DuplicateExists {
+			return errors.Errorf("Duplicate instance exists in the network for validator %#x", valRes.PublicKey)
+		}
+	}
```
Comment (prestonvanloon) [inline validator/client/validator.go:425]: Optional: check all of the validators and build an error with all of the duplicates found instead of returning on just the first one.

File: validator/client/validator.go [hunk 13]
```diff
@@ -369,6 +370,85 @@ func (v *validator) SlotDeadline(slot types.Slot) time.Time {
...
+	req := &ethpb.DoppelGangerRequest{ValidatorRequests: []*ethpb.DoppelGangerRequest_ValidatorRequest{}}
+	for _, pkey := range pubkeys {
+		attRec, err := v.db.AttestationHistoryForPubKey(ctx, pkey)
+		if err != nil {
+			return err
+		}
+		if len(attRec) == 0 {
+			// If no history exists we simply send in a zero
+			// value for the request type.
```
Comment (prestonvanloon) [inline validator/client/validator.go:396]:
```suggestion
			// If no history exists we simply send in a zero
			// value for the request epoch and root.
```

File: validator/client/validator.go [hunk 14]
```diff
@@ -369,6 +370,85 @@ func (v *validator) SlotDeadline(slot types.Slot) time.Time {
...
+// any duplicates active in the network.
+func (v *validator) CheckDoppelGanger(ctx context.Context) error {
+	if !featureconfig.Get().EnableDoppleGanger {
+		return nil
+	}
+	log.Info("Running Doppleganger Check")
+	pubkeys, err := v.keyManager.FetchValidatingPublicKeys(ctx)
+	if err != nil {
+		return err
+	}
```
Comment (prestonvanloon) [inline validator/client/validator.go:383]: Optional, but i think this is a nicer UX. I moved the log after the fetching validating keys so we could log the number of keys. Also removed some extra capitalization.

```suggestion
	pubkeys, err := v.keyManager.FetchValidatingPublicKeys(ctx)
	if err != nil {
		return err
	}
	log.WithField("keys", len(pubkeys)).Info("Running doppleganger check")
```

**Review by prestonvanloon — CHANGES_REQUESTED** [review]
A few comments. There are a few typos with Doppelganger vs Doppleganger. Correct is `Doppelganger`.

I have some concerns about the implied safety of this feature. Please correct me if these assumptions are wrong.

- It seems that if a validator was online and attesting, but did not increase in balance, then this implementation could return a false negative. 
- A quick restart, the check is ignored.

Why does a validator need to provide the last known attested epoch? 
Is it so that restarts would be quick? 
If this is the case and that's the trade off we want to make to prevent a 2 epoch mandatory delay on every restart, then we ought to log to the user "WARNING: Validator was running within the last 2 epochs and doppelganger detection will be skipped."

**Comment by potuz** [comment]
> I have some concerns about the implied safety of this feature. Please correct me if these assumptions are wrong.
> 
> * It seems that if a validator was online and attesting, but did not increase in balance, then this implementation could return a false negative.

This is correct. The odds of this happening are less than  10^{-5} already conditional to being running a doppelganger as I showed in the issue originally. We can do better than this by checking that the balance increased more than certain negative multiple of the `base_reward` but the gain is minimal (if any). Notice that attesting and getting a negative reward for 2 epochs can only happen if the validator was assigned two epochs in a row to a slot 0 and missed target and head in both votes in a row. This is incredibly rare. 

> * A quick restart, the check is ignored.
> 
Correct again

> Why does a validator need to provide the last known attested epoch?
> Is it so that restarts would be quick?
> If this is the case and that's the trade off we want to make to prevent a 2 epoch mandatory delay on every restart, then we ought to log to the user "WARNING: Validator was running within the last 2 epochs and doppelganger detection will be skipped."

I agree with the warning. 

But let me stress something that passes without notice in all the approaches waiting, like the one implemented in Lighthouse: the point is that the 2 epochs mandatory delay is completely useless in the case that this method fails to find a doppelganger, namely. In the rare event that the user starts two validators simultaneously, then if the mandatory two epochs delay is implemented, then both validators will wait two epochs, not find anything and then start happily validating at the same time and get slashed. 

So yes, this method will not find a doppelganger if both instances are launched together, but neither will waiting 2 epochs.

File: beacon-chain/rpc/validator/status.go (hunk 8 above)
Comment (potuz) [reply beacon-chain/rpc/validator/status.go:154]: There is nothing that can be really done for attestations you don't see. If a validator produces an attestation for a fork across epochs two epochs in a row this can happen, you can check for balances bigger than `-1/8 * base_reward` assuming perfect inclusion or even less, for example `-2 base_reward` would work even with minimal participation. But I claim it's not worth it. In such forked scenarios nothing can prevent you from being slashed. This feature is to prevent mistakes from users even in normal circumstances, as all slashings have been so far

File: beacon-chain/rpc/validator/status.go (hunk 6 above)
Comment (nisdas) [reply beacon-chain/rpc/validator/status.go:111]: This is to prevent an overflow `( currEpoch <= 1)` , in this case the doppelganger check will be skipped anyway as it is at the start of the chain.

File: beacon-chain/rpc/validator/status.go [hunk 15]
```diff
@@ -88,6 +88,105 @@ func (vs *Server) MultipleValidatorStatus(
...
+		Responses: []*ethpb.DoppelGangerResponse_ValidatorResponse{},
+	}
+	for _, v := range req.ValidatorRequests {
+		// If the validator's last recorded epoch was
+		// less than or equal to 2 epochs ago, this method will not
+		// be able to catch duplicates. This is due to how attestation
+		// inclusion works, where an attestation for the current epoch
+		// is able to included in the current or next epoch. Depending
+		// on which epoch it is included the balance change will be
+		// reflected in the following epoch.
```
Comment (prestonvanloon) [inline beacon-chain/rpc/validator/status.go:134]: This is much more clear! :+1:

File: shared/featureconfig/flags.go [hunk 16]
```diff
@@ -122,6 +122,12 @@ var (
 		Name:  "enable-optimized-balance-update",
 		Usage: "Enables the optimized method of updating validator balances.",
 	}
+	enableDoppelGangerProtection = &cli.BoolFlag{
+		Name: "enable-doppelganger",
+		Usage: "Enables the validator to perform a doppelganger check. (Warning): This is not " +
+			"a foolproof method to find duplicate instances in the network. Your validator will still be" +
+			" vulnerable if it is being run in unsafe configurations.",
```
Comment (prestonvanloon) [inline shared/featureconfig/flags.go:129]: Looks good!

File: validator/client/validator_test.go [hunk 17]
```diff
@@ -969,3 +970,233 @@ func TestService_ReceiveBlocks_SetHighest(t *testing.T) {
...
+					resp.ValidatorRequests = append(resp.ValidatorRequests, &ethpb.DoppelGangerRequest_ValidatorRequest{PublicKey: pkey[:], Epoch: att.Data.Target.Epoch, SignedRoot: rt[:]})
+				}
+				v := &validator{
+					validatorClient: client,
+					keyManager:      km,
+					db:              db,
+				}
+				client.EXPECT().CheckDoppelGanger(
+					gomock.Any(), // ctx
+					gomock.Any(), // request
```
Comment (prestonvanloon) [inline validator/client/validator_test.go:1010]: Could you add the actual requests for all test cases? This will add a bit more rigidity to these tests

Comment (nisdas) [reply validator/client/validator_test.go:1010]: So this is actually not trivial, I tried to do it previously, however the mock key manager actually uses a map underneath to store all the keys. When you request for all validating pubkeys, the ordering is non-deterministic, therefore any request mocked in here will be non deterministic too and basically only succeed 1/10 times. If needed I can create a new deterministic key manager to mock this.

File: validator/client/validator.go [hunk 18]
```diff
@@ -369,6 +370,94 @@ func (v *validator) SlotDeadline(slot types.Slot) time.Time {
...
+				})
+			continue
+		}
+		r := retrieveLatestRecord(attRec)
+		req.ValidatorRequests = append(req.ValidatorRequests,
+			&ethpb.DoppelGangerRequest_ValidatorRequest{
+				PublicKey:  r.PubKey[:],
+				Epoch:      r.Target,
+				SignedRoot: r.SigningRoot[:],
+			})
```
Comment (prestonvanloon) [inline validator/client/validator.go:411]: r.PubKey is always zero bytes. It is not set here: https://github.com/prysmaticlabs/prysm/blob/ca7b3127613bf9522d1b49dce4b1142f41702efe/validator/db/kv/attester_protection.go#L116-L119

File: validator/client/validator_test.go (hunk 17 above)
Comment (prestonvanloon) [reply validator/client/validator_test.go]: Added a custom matcher that will help with this.

File: validator/client/validator.go (hunk 18 above)
Comment (prestonvanloon) [reply validator/client/validator.go:414]: I've fixed this PR, but now other tests are failing due to this bug. I'll submit another PR to develop to fix the underlying issue with attester_protection.go AttestationHistoryForPubKey, then we can resync this PR with develop.

Comment (prestonvanloon) [reply validator/client/validator.go:414]: #9135 will fix this bug and round trip tests.

**Comment by prestonvanloon** [comment]
Blocked until https://github.com/prysmaticlabs/prysm/pull/9135 merges

File: validator/client/attest_test.go [hunk 19]
```diff
@@ -132,7 +133,7 @@ func TestAttestToBlockHead_AttestsCorrectly(t *testing.T) {
 	m.validatorClient.EXPECT().ProposeAttestation(
 		gomock.Any(), // ctx
 		gomock.AssignableToTypeOf(&ethpb.Attestation{}),
-	).Do(func(_ context.Context, att *ethpb.Attestation) {
+	).Do(func(_ context.Context, att *ethpb.Attestation, opts ...grpc.CallOption) {
```
Comment (prestonvanloon) [inline validator/client/attest_test.go:136]: This needed to be updated after updating gomock

File: validator/client/validator.go [hunk 20]
```diff
@@ -369,6 +370,97 @@ func (v *validator) SlotDeadline(slot types.Slot) time.Time {
...
+					PublicKey:  pkey[:],
+					Epoch:      0,
+					SignedRoot: make([]byte, 32),
+				})
+			continue
+		}
+		r := retrieveLatestRecord(attRec)
+		if pkey != r.PubKey {
+			return errors.New("attestation record mismatched public key")
+		}
```
Comment (prestonvanloon) [inline validator/client/validator.go:408]: Added this as a sanity check. It may not be necessary but it was some assurance that we received the correct record.

**Review by prestonvanloon — APPROVED** [review]

**Review by prestonvanloon — APPROVED** [review]
